content/blog/2019-11-13-strength-reduction-pass-in-llvm.md

```diff
@@ -0,0 +1,195 @@
+
+```shell
+# generate LLVM IR for original program 
+clang -c -emit-llvm -O0 -Xclang -disable-O0-optnone benchmark.c $EMBENCH_DIR/support/*.c -I/path/to/benchmark \
+-I$EMBENCH_DIR/support -DCPU_MHZ=1000
+
+for f in *.bc
+do
+  llvm-dis ${f} 
+done
+
+for f in *.ll
+do
+  # optimze with llvm opt
+  opt -S -load build/skeleton/libSkeletonPass.so -mem2reg -sr -dce ${f} -o opt_${f}
+  opt -S -load build/skeleton/libSkeletonPass.so -mem2reg -sr -dce ${f} -o opt_${f}.bc
+  llc -filetype=obj opt_${f}.bc; 
+done
+
+gcc *.o -lm; 
+time ./a.out
+```
+
+The LLVM pass is first compiled into a shared library, which is loaded into LLVM opt as a compiler pass (with -sr argument, which is the custom name of our pass). The experiment results is shown in the following table. To make a fair comparison, we also commented out all the content except the preprocessing and postprocessing part in the code of our strength reduction pass, compile it into a "dummy pass", and use the same commands to compile the benchmarks. The results of executing this version of the benchmarks is recorded as our baseline in the "Original" column of the table. 
+
+| Benchmark | Original (s) | Optimized (s) | Speedup |
+|:---------:|:------------:|:------------:|-----------|
+|aha-mont64|0.261|0.244| 1.070 |
+|crc32|0.598|0.598| 1 |
+|cubic|0.018|0.018| 1 |
+|edn|0.439|0.389| 1.129 |
+|huffbench|0.564|0.027| 20.889 |
+|matmult-int|0.628|0.579| 1.085 |
+|minver|0.067|0.078| 0.859 |
+|nbody|0.01|0.015| 0.667 |
+|nettle-aes|0.316|wrong| N/A |
+|nettle-sha256|0.348|0.364| 0.956 |
+|nsichneu|0.312|0.314| 0.994 |
+|picojpeg|0.545|stuck| N/A |
+|qrduino|1.183|1.136| 1.041 |
+|sglib-combined|0.612|wrong| N/A |
+|slre|0.729|wrong| N/A |
+|st|0.064|0.049| 1.306 |
+|statemate|0.372|0.401| 0.928 |
+|ud|0.55|0.538| 1.022 |
+|wikisort|0.215|0.276| 0.779 |
+|Geomean|||1.211|
+
+Except for a few designs, the average speedup is 1.211x over the original baseline program. Unfortunately, our pass does not optimize all programs correctly: `nettle-aes`, `sglib-combined`, and `slre` does not produce correct results after our optimization, while `picojpeg` does not terminate after applying our pass. We have not figured out the reason of these errors yet. Possible reasons are:
+- We failed to distinguish all the non-induction variables collected at the beginning of the pass in our analysis and optimization process, causing the program to behave incorrectly;
+- When inserting the update instructions, we probably inserted at incorrect positions or inserted multiple times. 
```

Comment:
Interesting! One other outlier in your dataset also probably bears closer examination: the 20x speedup for huffbench. Looks like [that app](https://github.com/embecosm/embench-beebs/blob/master/src/huffbench/libhuffbench.c) does lots of indirect array accesses, which could explain it.